src/polycap-description.c

```diff
@@ -0,0 +1,455 @@
+		exit(2);
+	}
+	fread(&seed, sizeof(unsigned long int), 1, random_device);
+	fclose(random_device);
+#endif
+	rng = polycap_rng_new(seed);
+
+
+	#pragma omp for nowait
+	for(j=0; j < icount; j++){
+		// Obtain photon start coordinates
+		n_shells = round(sqrt(12. * description->n_cap - 3.)/6.-0.5);
+		if(n_shells == 0.){ //monocapillary case
+			r = polycap_rng_uniform(rng);
+			start_coords.x = (2.*r-1) * description->profile->cap[0];
+			r = polycap_rng_uniform(rng);
+			start_coords.y = (2.*r-1) * description->profile->cap[0];
+			start_coords.z = 0.;
+		} else { // polycapillary case
+			// select random coordinates, check whether they are inside polycap boundary
+			boundary_check = -1;
+			do{
+				r = polycap_rng_uniform(rng);
+				start_coords.x = (2.*r-1) * description->profile->ext[0];
+				r = polycap_rng_uniform(rng);
+				start_coords.y = (2.*r-1) * description->profile->ext[0];
+				start_coords.z = 0.;
+				boundary_check = polycap_photon_within_pc_boundary(description->profile->ext[0], start_coords);
+			} while(boundary_check == -1);
+		}
+	
+		// Obtain point from source as photon origin, determining photon start_direction
+		r = polycap_rng_uniform(rng);
+		src_rad_x = description->src_x * sqrt(fabs(r)); ////sqrt to simulate source intensity distribution (originally src_x * r/sqrt(r) )
+		r = polycap_rng_uniform(rng);
+		src_rad_y = description->src_y * sqrt(fabs(r)); ////sqrt to simulate source intensity distribution
+		r = polycap_rng_uniform(rng);
+		phi = 2.0*M_PI*fabs(r);
+		src_start_x = src_rad_x * cos(phi) + description->src_shiftx;
+		src_start_y = src_rad_y * sin(phi) + description->src_shifty;
+		if((description->src_sigx * description->src_sigy) < 1.e-20){ //uniform distribution over PC entrance
+			r = polycap_rng_uniform(rng);
+			pc_rad = description->profile->ext[0] * sqrt(fabs(r));
+			r = polycap_rng_uniform(rng);
+			phi = 2.0*M_PI*fabs(r);
+			pc_x = pc_rad * cos(phi) + start_coords.x;
+			pc_y = pc_rad * sin(phi) + start_coords.y;
+			start_direction.x = pc_x - src_start_x;
+			start_direction.y = pc_y - src_start_y;
+			start_direction.z = description->d_source;
+		} else { //non-uniform distribution, direction vector is within +- sigx
+			r = polycap_rng_uniform(rng);
+			start_direction.x = description->src_sigx * (1.-2.*fabs(r));
+			r = polycap_rng_uniform(rng);
+			start_direction.y = description->src_sigy * (1.-2.*fabs(r));
+			start_direction.z = 1.;
+		}
+		polycap_norm(&start_direction);
+
+		// Create photon structure
+		photon = polycap_photon_new(rng, start_coords, start_direction, start_electric_vector, n_energies, energies);
+
+		// Launch photon
+		istart++; //Here all photons that started, also enter the polycapillary
+		photon->i_refl = 0; //set reflections to 0
+		iesc = polycap_photon_launch(photon, description);
+			//if iesc == -1 here a new photon should be simulated/started.
+			//	essentially do j-1 as this will have same effect
+		if(iesc == -1) j--;
```

Comment:
Dit lijkt mij iets bijzonder ondeugend: ik zou zeker niet aan de OpenMP private variable `j` komen aangezien dit niet zal doen wat jij denkt...

Wil je dat het volgend foton gesimuleerd wordt zonder dat de rest van de for loop block wordt uitgevoerd voor het huidige foton? Gebruik dan gewoon `continue;`, vooraf te gaan door free'en van geheugen waar nodig.

Reply:
Ja, was al bang dat het niet koosjer was :) Ging het nog vragen, maar vergeten.

Daar moet hij eigenlijk nieuwe photon start coordinaten (en richting etc) geven, maar dus zonder j te incrementen. Dit komt omdat de OpenMP for loop eigenlijk loopt over 5000 gedetecteerde fotonen, terwijl als iesc == -1 op dat moment het wil zeggen dat het foton niet uit het capillair is geraakt, en dus gedetecteerd. Dus we willen een nieuw foton simuleren, zonder dat het increment stijgt.
Kan het perfect doen met een do..while loop, zoals in eerdere versies, maar je was daar niet echt fan van meen ik te herinneren.

Reply:
In dat geval best terug te gaan naar een do..while loop.